orm/serial_bucket.go

```diff
@@ -0,0 +1,384 @@
+package orm
+
+import (
+	"reflect"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+)
+
+// SerialModel is implemented by any entity that can be stored using SerialModelBucket.
+// GetID/SetID are used to store and access the Key.
+// The ID is always set to nil before serializing and storing the Value.
+type SerialModel interface {
+	weave.Persistent
+	Validate() error
+	GetID() []byte
+	SetID([]byte) error
+}
+
+// SerialModelSlicePtr represents a pointer to a slice of SerialModels. Think of it as
+// *[]SerialModel Because of Go type system, using []SerialModel would not work for us.
+// Instead we use a placeholder type and the validation is done during the
+// runtime.
+type SerialModelSlicePtr interface{}
+
+// SerialModelBucket is implemented by buckets that operates on SerialModels rather than
+// Objects.
+type SerialModelBucket interface {
+	// One query the database for a single SerialModel instance. Lookup is done
+	// by the primary index key. Result is loaded into given destination
+	// SerialModel.
+	// This method returns ErrNotFound if the entity does not exist in the
+	// database.
+	// If given SerialModel type cannot be used to contain stored entity, ErrType
+	// is returned.
+	One(db weave.ReadOnlyKVStore, key []byte, dest SerialModel) error
+
+	// PrefixScan will scan for all SerialModels with a primary key (ID)
+	// that begins with the given prefix.
+	// The function returns a (possibly empty) iterator, which can
+	// load each SerialModel as it arrives.
+	// If reverse is true, iterates in descending order (highest value first),
+	// otherwise, it iterates in ascending order.
+	PrefixScan(db weave.ReadOnlyKVStore, prefix []byte, reverse bool) (SerialModelIterator, error)
+
+	// ByIndex returns all objects that secondary index with given name and
+	// given key. Main index is always unique but secondary indexes can
+	// return more than one value for the same key.
+	// All matching entities are appended to given destination slice. If no
+	// result was found, no error is returned and destination slice is not
+	// modified.
+	ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest SerialModelSlicePtr) error
+
+	// IndexScan does a PrefixScan, but on the named index. This would let us eg. load all counters
+	// in order of their count. Or easily find the lowest or highest count.
+	IndexScan(db weave.ReadOnlyKVStore, indexName string, prefix []byte, reverse bool) (SerialModelIterator, error)
+
+	// Create saves given SerialModel in the database. Before inserting into
+	// database, SerialModel is validated using its Validate method.
+	// ID field must be unset so auto incremented ID is generated.
+	Create(db weave.KVStore, m SerialModel) error
+
+	// Upsert creates given SerialModel in the database or upserts of given SerialModel
```

Comment:
Having those atomic operations would be valid when many clients are accessing the database concurrently. In our case, we run a single threaded program, each event being processed within a transaction, and we use a local database. 
I think in our case, it is fine to call `Has` or `Get` before storing an entity to ensure consistency with put operation. As a consequence we can have a small and simple interface for accessing the data. I believe this is a good thing, because it is easier to remember what function does what.

In my opinion, `Create`, `Upsert` can be reduced to a single method. You can check the intent of the user by investigating ID field.

Reply:
I like single method approach :)